### PR TITLE
Armour Refactor

### PR DESCRIPTION
**Pull request by Edan52**
## What Does This PR Do
Applies the idea proposed in #18521

Refactors the armour calculation & modifies some of the existing armour values to be more in line with these changes. This does make the code a bit harder to understand initially due to the higher complexity of the calculation, I do intend to do what I can to explain what's happening (updating the armour wiki, providing tools).

This means that individual armour pieces should remain effectively unchanged, within a couple percent.

Armour code is scattered in many places, so please feel free to comment or ping me on discord with anything I might have missed.

Armour penetration now comes in two forms, percentage and raw. Percentage is calculated first and removes a percentage of your total armour, and then the raw armour is taken away. For example 75 armour hit by a 30% armour penetration followed by 20 raw penetration would look like this:

75 * 0.7 = 52.5 armour
52.5 - 20 = 32.5 final armour.

Percentage is not implemented yet, but the framework is there for items and simple mobs and it'd be good to go over armour penetration values either in this PR or another one as there are some potential issues that could rise if they remain unchanged.

Shield code has also been cleaned up and now prioritises the item you're holding that gives the highest block chance, rather than just your left hand and then right hand.

## What does this mean for your every day round?

Security jumpsuit + sec armour now results in 25 melee armour which is a 33% **melee** damage reduction, rather than the 40% it was before. The rest of the damage types remain effectively unchanged.

Tactical turtleneck + blood-red hardsuit results in 45 melee armour, which is about a 47% reduction in melee damage, whereas before it would have been 50%

Tactical turtleneck + elite syndicate hardsuit results in 80 melee armour, which is about a 62% reduction in melee damage, where it would've been 70% before.

Mining armour stacking has been nerfed quite a bit, you won't be able to reach much more than 70% reduction now.

All I have done thus far is change the values to be the same as they were before, so some common combinations of armour have most likely also been nerfed. If requested, I could buff these to be more equal to what they were before however I'd like to keep the refactor atomic to that.

Armour penetrations have been tweaked by me and @hal9000PR to be around about as effective in combat as they were before, there is no simple calculation to convert the values so we've had to consider what weapons are used in which situations against which armour and assign values accordingly.

# Please leave comments relevant to the code on this PR, talk about the idea on the discussion thread.

## Why It's Good For The Game
See reasoning on the project discussion.

## Changelog
:cl: CornMyCob
tweak: Armour now uses a new calculation, stacking lots of armour is now less effective.
tweak: Shields now prioritise the item you're holding that has the highest block chance.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/human/species/_species.dm [hunk 1]
```diff
@@ -375,7 +375,7 @@
 
 	switch(damagetype)
 		if(BRUTE)
-			var/damage_amount = damage * hit_percent * brute_mod * H.physiology.brute_mod
+			var/damage_amount = (damage/(1+(blocked/ARMOUR_CONSTANT))) * brute_mod * H.physiology.brute_mod
```
Comment (hal9000PR) [inline code/modules/mob/living/carbon/human/species/_species.dm:378]: this line is written multiple times just with different mods.
might it be good to make a define like
```
#define ARMOUR_EQUATION(damage, armour, modifier) (damage / (1 + (armour / ARMOUR_CONSTANT)) * modifier
```
and then pass the damage, blocked and mod1 * mod2 into the params.

**Comment by Qwertytoforty** [comment]
I am not sure I am a fan of this formula, that makes armor under 50 stronger and over 50 weaker. I feel like addressing the sources of stacking, sec jumpsuit, armored accessories, and bone bracers, may be better than this. This means we can keep the armor values simple to read, with reading 20 melee being a 20% reduction, vs actually being ~29% reduction. Also keeps it more consistent with how the rest of the armor works in the game. Alternatively, we make stacking armor items multiplicitive, vs additive.

**Comment by Rurik123** [comment]
I'm an idiot when it comes to this sort of thing, so feel free to disregard/correct this if I've made a mistake.

If I'm reading this right, the armor value for anything above 50 is nerfed, even if it is not relying on stacking armor. Like Elite Syndicate HS being lowered from 60 soley cause it stands alone above 50, not cause of any additional armor.

Wouldn't it be superior if the formula made the largest armor value flat, and instead adjusted the efficiency of the smaller stacking armor values? This way the established armor value of inquis/ERT/syndicate/etc armor is kept, as well as the armor stacking problem being solved?

Apologies if I misread or misunderstand something.

**Comment by SpringSkipper** [comment]
The current armor system is entirely fine except for the additive stacking.

If you change the formula then you're taking it on yourself to change the armor amount of every armor item in the game. That seems like a hard requirement to me. Otherwise you're blindly nerfing and buffing a lot of items without regard to the game effect those changes will have.

Changing stacking to multiplicative would work just as well at fixing the stacking mining armor problem, plus it would remove the requirement to readjust every armor value in the whole game.

File: code/modules/mob/living/carbon/human/human.dm [hunk 2]
```diff
@@ -222,7 +222,7 @@
 
 	var/brute_loss = 0
 	var/burn_loss = 0
-	var/bomb_armor = getarmor(null, BOMB)
+	var/bomb_armor = 100 - (100/(1+(getarmor(null, BOMB)/50)))
```
Comment (Edan52) [inline code/modules/mob/living/carbon/human/human.dm:225]: Gonna change this calculation to a define (and add a couple more defines for people to use) and apply it to a couple more cases I found after work.

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 3]
```diff
@@ -216,26 +215,32 @@ emp_act
...
+	else if(r_block_chance > 0 && r_hand.hit_reaction(src, AM, attack_text, r_block_chance, damage, attack_type))
+		return TRUE
+
+	if(wear_suit && wear_suit.hit_reaction(src, AM, attack_text, 0, damage, attack_type))
+		return TRUE
+
+	if(w_uniform && w_uniform.hit_reaction(src, AM, attack_text, 0, damage, attack_type))
+		return TRUE
+
+	return FALSE
```
Comment (hal9000PR) [inline code/modules/mob/living/carbon/human/human_defense.dm:241]: much clearner. good stuff.

File: code/modules/mob/living/living_defense.dm [hunk 4]
```diff
@@ -10,27 +10,28 @@
...
 		else
 			to_chat(src, "<span class='userdanger'>Your armor absorbs the blow!</span>")
-	else if(armor > 0)
-		if(soften_text)
-			to_chat(src, "<span class='userdanger'>[soften_text]</span>")
+		return armor
+
+	if(armor > 0)
+		if(armour_penetration_flat > 0 || armour_penetration_percentage > 0)
+			armor = max(0, (armor * ((100-armour_penetration_percentage) / 100)) - armour_penetration_flat)
```
Comment (hal9000PR) [inline code/modules/mob/living/living_defense.dm:25]:
```suggestion
			armor = max(0, (armor * ((100 - armour_penetration_percentage) / 100)) - armour_penetration_flat)
```

File: code/modules/mob/living/simple_animal/hostile/megafauna/ancient_robot.dm [hunk 5]
```diff
@@ -247,7 +247,7 @@ Difficulty: Hard
 				L.visible_message("<span class='danger'>[src] slams into [L]!</span>", "<span class='userdanger'>[src] tramples you into the ground!</span>")
 				forceMove(get_turf(L))
 				var/limb_to_hit = L.get_organ(pick(BODY_ZONE_HEAD, BODY_ZONE_CHEST, BODY_ZONE_R_ARM, BODY_ZONE_L_ARM, BODY_ZONE_R_LEG, BODY_ZONE_L_LEG))
-				L.apply_damage(25, BRUTE, limb_to_hit, L.run_armor_check(limb_to_hit, MELEE, null, null, armour_penetration))
+				L.apply_damage(25, BRUTE, limb_to_hit, L.run_armor_check(limb_to_hit, MELEE, null, null, armour_penetration_flat))
```
Comment (hal9000PR) [inline code/modules/mob/living/simple_animal/hostile/megafauna/ancient_robot.dm:250]: dont you want to pass percentage armour pen into this?

File: code/modules/mob/living/simple_animal/hostile/megafauna/bubblegum.dm [hunk 6]
```diff
@@ -253,7 +253,7 @@ Difficulty: Hard
 			to_chat(L, "<span class='userdanger'>[src] rends you!</span>")
 			playsound(T, attack_sound, 100, TRUE, -1)
 			var/limb_to_hit = L.get_organ(pick(BODY_ZONE_HEAD, BODY_ZONE_CHEST, BODY_ZONE_R_ARM, BODY_ZONE_L_ARM, BODY_ZONE_R_LEG, BODY_ZONE_L_LEG))
-			L.apply_damage(10, BRUTE, limb_to_hit, L.run_armor_check(limb_to_hit, MELEE, null, null, armour_penetration))
+			L.apply_damage(10, BRUTE, limb_to_hit, L.run_armor_check(limb_to_hit, MELEE, null, null, armour_penetration_flat))
```
Comment (hal9000PR) [inline code/modules/mob/living/simple_animal/hostile/megafauna/bubblegum.dm:256]: dont you want to pass percentage armour pen into this?

File: code/modules/mob/living/simple_animal/hostile/megafauna/colossus.dm [hunk 7]
```diff
@@ -253,7 +253,7 @@ Difficulty: Very Hard
 	name ="death bolt"
 	icon_state= "chronobolt"
 	damage = 25
-	armour_penetration = 100
+	armour_penetration_flat = INFINITY
```
Comment (hal9000PR) [inline code/modules/mob/living/simple_animal/hostile/megafauna/colossus.dm:256]: armour pen percentage of 100

File: code/modules/mob/living/simple_animal/hostile/megafauna/ancient_robot.dm [hunk 8]
```diff
@@ -630,7 +630,7 @@ Difficulty: Hard
 			L.visible_message("<span class='danger'>[src] slams into [L]!</span>", "<span class='userdanger'>[src] tramples you into the ground!</span>")
 			forceMove(get_turf(L))
 			var/limb_to_hit = L.get_organ(pick(BODY_ZONE_HEAD, BODY_ZONE_CHEST, BODY_ZONE_R_ARM, BODY_ZONE_L_ARM, BODY_ZONE_R_LEG, BODY_ZONE_L_LEG))
-			L.apply_damage(12.5, BRUTE, limb_to_hit, L.run_armor_check(limb_to_hit, MELEE, null, null, armour_penetration))
+			L.apply_damage(12.5, BRUTE, limb_to_hit, L.run_armor_check(limb_to_hit, MELEE, null, null, armour_penetration_flat))
```
Comment (hal9000PR) [inline code/modules/mob/living/simple_animal/hostile/megafauna/ancient_robot.dm:633]: same here

File: code/modules/mob/living/simple_animal/hostile/syndicate.dm [hunk 9]
```diff
@@ -39,7 +39,7 @@
 	icon_living = "syndicate_sword"
 	attacktext = "slashes"
 	attack_sound = 'sound/weapons/bladeslice.ogg'
-	armour_penetration = 28
+	armour_penetration_flat = 28
```
Comment (hal9000PR) [inline code/modules/mob/living/simple_animal/hostile/syndicate.dm:42]: why the fuck is this 28 armour pen??

File: code/modules/mob/living/simple_animal/simple_animal.dm [hunk 10]
```diff
@@ -56,7 +56,8 @@
 	var/melee_damage_lower = 0
 	var/melee_damage_upper = 0
 	var/obj_damage = 0 //how much damage this simple animal does to objects, if any
-	var/armour_penetration = 0 //How much armour they ignore, as a flat reduction from the targets armour value
+	var/armour_penetration_flat = 0 //How much armour they ignore, as a flat reduction from the targets armour value
+	var/armour_penetration_percentage = 0 //How much armour they ignore, as a percentage of their total armour
```
Comment (hal9000PR) [inline code/modules/mob/living/simple_animal/simple_animal.dm:60]:
```suggestion
	/// How much armour they ignore, as a flat reduction from the targets armour value. calculated after percentage armour pen
	var/armour_penetration_flat = 0
	/// How much armour they ignore, as a percentage of their total armour. calculated before flat armour pen.
	var/armour_penetration_percentage = 0
```

File: code/modules/projectiles/projectile/magic.dm [hunk 11]
```diff
@@ -4,7 +4,7 @@
 	damage = 0
 	damage_type = OXY
 	nodamage = 1
-	armour_penetration = 100
+	armour_penetration_flat = INFINITY
```
Comment (hal9000PR) [inline code/modules/projectiles/projectile/magic.dm:7]: percent armour pen

**Review by hal9000PR — CHANGES_REQUESTED** [review]

File: code/datums/weather/weather_types/radiation_storm.dm [hunk 12]
```diff
@@ -38,7 +38,7 @@
 		if(ishuman(L))
 			var/mob/living/carbon/human/H = L
 			if(!HAS_TRAIT(H, TRAIT_RADIMMUNE))
-				if(prob(max(0, 100 - resist)))
+				if(prob(max(0, 100 - ARMOUR_VALUE_TO_PERCENTAGE(resist))))
```
[the remark this answers is not in the document]
Comment (Edan52) [reply code/datums/weather/weather_types/radiation_storm.dm:41]: done

File: code/__DEFINES/armour.dm [hunk 13]
```diff
@@ -0,0 +1,6 @@
+// Armour defines for the new armour system.
+
+#define ARMOUR_CONSTANT 50
+#define ARMOUR_EQUATION(damage, armour, modifier) ((damage / (1 + (armour / ARMOUR_CONSTANT))) * modifier)
+#define ARMOUR_PERCENTAGE_TO_VALUE(percentage) ((5000 / (100 - percentage)) - 50)
```
[the remark this answers is not in the document]
Comment (Edan52) [reply code/__DEFINES/armour.dm]: added a check

File: code/game/gamemodes/cult/cult_items.dm [hunk 14]
```diff
@@ -200,7 +200,7 @@
 	flags_inv = HIDEJUMPSUIT
 	allowed = list(/obj/item/tome, /obj/item/melee/cultblade)
 	body_parts_covered = UPPER_TORSO|LOWER_TORSO|LEGS|ARMS
-	armor = list(MELEE = -50, BULLET = -50, LASER = -50, ENERGY = -50, BOMB = -50, BIO = -50, RAD = -50, FIRE = 0, ACID = 0)
+	armor = list(MELEE = -25, BULLET = -25, LASER = -25, ENERGY = -25, BOMB = -25, BIO = -25, RAD = -25, FIRE = 0, ACID = 0)
```
[the remark this answers is not in the document]
Comment (Edan52) [reply code/game/gamemodes/cult/cult_items.dm:203]: added a check, so this should never occur

File: code/game/objects/items.dm [hunk 15]
```diff
@@ -66,7 +66,8 @@ GLOBAL_DATUM_INIT(welding_sparks, /mutable_appearance, mutable_appearance('icons
 	var/permeability_coefficient = 1 // for chemicals/diseases
 	var/siemens_coefficient = 1 // for electrical admittance/conductance (electrocution checks and shit)
 	var/slowdown = 0 // How much clothing is slowing you down. Negative values speeds you up
-	var/armour_penetration = 0 //percentage of armour effectiveness to remove
+	var/armour_penetration_flat = 0 //flat armour reduction
+	var/armour_penetration_percentage = 0 //percentage armour reduction
```
[the remark this answers is not in the document]
Comment (Edan52) [reply code/game/objects/items.dm]: applied

File: code/game/objects/items/weapons/batons.dm [hunk 16]
```diff
@@ -112,7 +112,7 @@
   * * user - The attacking user
   */
 /obj/item/melee/classic_baton/proc/on_non_silicon_stun(mob/living/target, mob/living/user)
-	var/armour = target.run_armor_check("chest", armour_penetration = stamina_armour_pen) // returns a % of their chest melee armour
+	var/armour = target.run_armor_check("chest", armour_penetration_flat = stamina_armour_pen) // returns a % of their chest melee armour
```
[the remark this answers is not in the document]
Comment (Edan52) [reply code/game/objects/items/weapons/batons.dm]: done

File: code/game/objects/items/weapons/batons.dm [hunk 17]
```diff
@@ -112,7 +112,7 @@
...
 	var/factor = (100 - armour) / 100 // converts the % into a decimal
 	target.adjustStaminaLoss(stamina_damage * factor)
```
[the remark this answers is not in the document]
Comment (Edan52) [reply code/game/objects/items/weapons/batons.dm]: added a human check, uses the percentage calculation & is the same as before.

File: code/game/objects/items/weapons/melee/energy.dm [hunk 18]
```diff
@@ -99,7 +99,7 @@
 	w_class_on = WEIGHT_CLASS_HUGE
 	hitsound = 'sound/weapons/bladeslice.ogg'
 	flags = CONDUCT
-	armour_penetration = 100
+	armour_penetration_flat = INFINITY
```
[the remark this answers is not in the document]
Comment (Edan52) [reply code/game/objects/items/weapons/melee/energy.dm]: done

File: code/game/objects/items/weapons/staff.dm [hunk 19]
```diff
@@ -8,7 +8,7 @@
 	throw_speed = 1
 	throw_range = 5
 	w_class = WEIGHT_CLASS_SMALL
-	armour_penetration = 100
+	armour_penetration_flat = INFINITY
```
[the remark this answers is not in the document]
Comment (Edan52) [reply code/game/objects/items/weapons/staff.dm]: done

File: code/game/objects/items/weapons/twohanded.dm [hunk 20]
```diff
@@ -576,7 +576,7 @@
 /obj/item/twohanded/required/chainsaw/doomslayer
 	name = "OOOH BABY"
 	desc = "<span class='warning'>VRRRRRRR!!!</span>"
-	armour_penetration = 100
+	armour_penetration_flat = INFINITY
```
[the remark this answers is not in the document]
Comment (Edan52) [reply code/game/objects/items/weapons/twohanded.dm]: done

File: code/__DEFINES/armour.dm [hunk 21]
```diff
@@ -0,0 +1,6 @@
...
+#define ARMOUR_VALUE_TO_PERCENTAGE(value) (100 - (100 / (1 + (value / ARMOUR_CONSTANT))))
```
[the remark this answers is not in the document]
Comment (Edan52) [reply code/__DEFINES/armour.dm]: added a check

File: code/modules/mob/living/simple_animal/hostile/syndicate.dm (hunk 9 above)
Comment (Edan52) [reply code/modules/mob/living/simple_animal/hostile/syndicate.dm]: changed it to 30 :)

File: code/datums/weather/weather_types/radiation_storm.dm [hunk 22]
```diff
@@ -37,7 +37,7 @@
 	if(prob(40))
 		if(ishuman(L))
 			var/mob/living/carbon/human/H = L
-			if(!HAS_TRAIT(H, TRAIT_RADIMMUNE))
+			if(!HAS_TRAIT(H, TRAIT_RADIMMUNE) || resist != INFINITY)
```
Comment (hal9000PR) [inline code/datums/weather/weather_types/radiation_storm.dm:40]: shouldnt this be an `&&` ?

File: code/modules/mob/living/simple_animal/hostile/megafauna/ancient_robot.dm (hunk 8 above)
Comment (hal9000PR) [reply code/modules/mob/living/simple_animal/hostile/megafauna/ancient_robot.dm:633]: and you missed this one.

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 23]
```diff
@@ -216,26 +215,32 @@ emp_act
 
 //End Here
 
-/mob/living/carbon/human/proc/check_shields(atom/AM, damage, attack_text = "the attack", attack_type = MELEE_ATTACK, armour_penetration = 0)
+#define BLOCK_CHANCE_CALCULATION(hand_block_chance, armour_penetration_flat, armour_penetration_percentage, block_chance_modifier) clamp((hand_block_chance * ((100-armour_penetration_percentage) / 100)) - armour_penetration_flat + block_chance_modifier, 0, 100)
```
Comment (AffectedArc07) [inline code/modules/mob/living/carbon/human/human_defense.dm:218]: Should this go in the `armour.dm` defines file?

**Review by hal9000PR — APPROVED** [review]

**Pull request merged**